### DISCUSSION TITLE
GET requests failing on https://kanji.koohii.com/main

### DISCUSSION THREAD

**Discussion by JFen10** [fabd/kanji-koohii#227] [3. Bugs]
After completing my restudy kanji today I am unable to load the /main page aside from as pictured below both on my iPad (Safari) and on my Macbook Pro (Chrome).

<img width="1440" alt="Screen Shot 2021-12-20 at 11 47 20 pm" src="https://user-images.githubusercontent.com/83437902/146769581-9466f0e9-128a-43f3-a502-37c431bd1cc6.png">

**Comment by fabd** [comment]
Hi @JFen10 ... have you started a new Kanji Review or SRS session yet? Could be that you started a review as I updated the site.. sorry for the trouble :/

Please let me know if the issue persists.

Another less likely thing is due to changes in the way Javascript is served; it may not pick up new files. If you do a full refresh in the browser hopefully it should work.

**Reply by JFen10** [reply]
I was working on this review for around a hour and a half and finished up about 25 minutes ago. I am unable to start another review as it just ends up landing on the page I posted above.

I tried both a hard refresh and empty cache and hard reload without any luck. I am also seeing the same issue on my phone too so I believe it isn't relating to my browser.

The number of "Study" cards is much lower than it should have been after completing my last session too. I can only see 10 when there should have probably been over 100 there.

<img width="810" alt="Screen Shot 2021-12-21 at 12 11 43 am" src="https://user-images.githubusercontent.com/83437902/146772334-e5c41b38-6a55-49cd-b59b-77a059576b3a.png">

**Reply by fabd** [reply]
Hmm thank you for the detailed reply.

I don't think there is a problem with data, knowing of the server syncs ratings.

My hunch is there is some problematic "half state" caused by the update if you started a review -- something left in the session, I can't think of right now what would cause it but I'll dig into it later today.

Worst case, see if you can NOT visit the site at all for at least 30 mins, this will end the temporary session on the server. If you can access the page afterwards this will give us a clue. 

If not in the meantime I'll dig into the code, it's not obvious for me because I am not seeing the error on my accounts on live site.

**Reply by JFen10** [reply]
Thanks for your quick response on this one too. It is past midnight in Australia now so I'll head to bed and check again once I wake up in the morning. I'll let you know if there is any change then.

**Reply by fabd** [reply]
I can't reproduce the error for now.

I am also have to "hit the sack" before I can fix this bug.

In the meantime once you see this could you send me your exact username, here or via the Contact page on the site if you like. Looking at the flashcard data could give me a clue.

It appears there was definitely incorrect handling of ratings. However in my test scenario the server should have interrupted your review with error 500. Then after refreshing page you'd be able to resume having to repeat maybe the last 10 cards (the client syncs every 10 cards -- it doesn't sync an entire review at once).

Worst case it will have rescheduled cards in each individual box, without actually changing the box, explaining why you don't see the  Restudy cards you expected to see at end of review > they stayed in whatever Leitner box they were.

Hard to say for now.

**Reply by c4b3z0n** [reply]
Hello I'm having the same issue.
I had started a review session at least one day before the update, but I tend to not end the session and leave the tab open.
Today I did continue my session in the same tab, and after a couple of reviews I ended the session. I did notice at some point during the reviews that suddenly the text-color of the keywords was not the same as usual, I guess it had to do with the update.
After clicking End, it went to the results page, but it was kind of broken. Like the rates on top were showing NaN%
Also the table with the kanji I had reviewed was kind of broken, I don't remember exactly how but there was some missing info on some columns, though the kanji appeared to be the ones I had reviewed.
I also think I had reviewed 21 kanji, though when I go to the /main, it says "19 reviews today"
On a side note, I also did swap the keywords for two kanji, one of which I was reviewing that session, don't know if this could had also led to the problem.
My username is c4b3z0n
Full browser refresh didn't fix the problem.

**Reply by fabd** [reply]
Hi @c4b3z0n see my reply below (not sure if you get email notified for the at @ mention alone).

**Comment by fabd** [comment]
Quick update, before lunch break.

Running some queries confirms my hunch. The SRS chart is stuck because the cards have been mistakenly updated. 🥵

Specifically the column `leitnerbox` in the database is 0 -- which the code doesn't account for.

I thought a SQL query would fail if the value has no explicit default and "cannot be null"... but alas. One learns everyday...

This query shouldn't have gone through either. It's all my mistake. It's all the more silly in that I did account for the old ratings for the _Kanji Ryokucha_ client, for which a `normalizeOldRatings` function is called. The function as the name implies, converts old ratings (sent be the iOS client), to newer ones... However I completely overlooked this situation for desktop clients, with a review *in progress* as I updated the website.

So what happened is the old Javascript in the client would send old ratings, while the updated server code didn't recognize them and did not properly reject them.

As a side effect because leitnerbox 0 is not valid, it rescheduled cards at a 0 interval, so all the cards affected have `expiredate` the same as the review time, yesterday (so for the site's logic it's as if they are all DUE).

**Thankfully because leitnerbox 0 is unexpected , I can find exactly who has been affected.**

A preliminary query shows exactly 9 users, including @c4b3z0n and @JFen10 :

```
+----------+--------+---------------------+
| count(*) | userid | lastreview          |
+----------+--------+---------------------+
|       12 | ****** | 2021-12-20 13:36:52 |
|        9 | ****** | 2021-12-20 19:37:58 |
|       19 | ****** | 2021-12-20 13:38:43 |
|       29 | ****** | 2021-12-20 12:51:21 |
|       11 | ****** | 2021-12-20 05:36:31 |
|       17 | ****** | 2021-12-20 05:52:03 |
|      221 | ****** | 2021-12-20 20:54:07 |
|       24 | ****** | 2021-12-20 07:19:48 |
|       20 | ****** | 2021-12-20 17:26:02 |
+----------+--------+---------------------+
```

The lastreview timestamps confirm those are all reviews from yesterday.

It shows that unfortunately @JFen10 is the most affected with 221 cards ( @c4b3z0n  has 19 cards affected ).

### Solutions

In order to fix this definitely I'll need to update the leitnerbox.

For @JFen10 what we could do is since you said you expected 100+ cards to be in the failed pile, we can move them to the failed pile. For failed cards there will be little harm done -- however for those cards you answered HARD/YES/EAST this is not ideal. If you want I could try to set the box , based on the success/fail ratio, to try and reduce the amount of cards sent to the failed pile. Or I can put them all back into the Failed pile , or I can put all of them with a fixed interval if you'd like to go through them again in 1 day, or 1 week etc. Since this could be long I propose to figure out the more convenient fix for you in the email conversation.

I'm really sorry for the trouble caused. 😓

I learned a few things here.

- one is that I should not have removed some "foolproof" code. I thought maybe I was a bad developer for having a bunch of conditionals in some places where I thought '"this situation should not happen anyway"... but for critical areas such as rating the cards, I should have a check in place, to ignore any unexpected rating.
- on the plus side, I did actually write tests for the first time, as a part of this udpate, I have a bunch of tests for the LeitnerSRS class, that verify all ratings move cards to the correct box and with expected range of intervals -- unfortunaly those tests only account for known ratings.
- SQL query doesn't fail even if the TABLE does not give explicit defaults -- I thought SQL queries would fail, as the query I believe does something like `... leitnerbox=''` and it fails on my local env, but the production server is more lax
- I thought using leitnerbox starting at 1 was a bad design because it makes the logic more confusing (since most comparisions are with values that start at 0) ... but this experience showed me it was a good idea after all, since I am able to find the invalid records exactly -- so I will keep this design

@c4b3z0n please also send me an email and let me know what you'd like to do with the 19 cards affected. I can move them to failed pile (simplest), or make them all due in 1 day, or N days..whichever is convenient. Once I apply this change, in theory the SRS chart will show up and everything works as expected.

**Reply by fabd** [reply]
PS: thankfully only `leitnerbox` and `expiredate` (due date) are incorrect, here is a sample of affected cards showing the other fields are left untouched:

```
+--------+--------+---------------------+---------------------+------------+--------------+------------+--------------+--------------+
| userid | ucs_id | created_on          | lastreview          | expiredate | totalreviews | leitnerbox | failurecount | successcount |
+--------+--------+---------------------+---------------------+------------+--------------+------------+--------------+--------------+
| ****** |  20195 | 2021-06-08 04:10:49 | 2021-12-20 20:42:38 | 2021-12-20 |           22 |          0 |            7 |           15 |
| ****** |  20208 | 2021-09-30 06:58:47 | 2021-12-20 21:08:14 | 2021-12-20 |           20 |          0 |           10 |           10 |
| ****** |  20365 | 2021-06-06 08:22:48 | 2021-12-20 21:02:45 | 2021-12-20 |           47 |          0 |           27 |           20 |
| ****** |  20498 | 2021-06-07 05:13:32 | 2021-12-20 20:55:46 | 2021-12-20 |           51 |          0 |           34 |           17 |
| ****** |  20511 | 2021-07-05 06:32:30 | 2021-12-20 21:14:40 | 2021-12-20 |           39 |          0 |           21 |           18 |
| ****** |  20625 | 2021-06-07 05:13:32 | 2021-12-20 20:47:27 | 2021-12-20 |           78 |          0 |           47 |           31 |
| ****** |  20693 | 2021-10-19 05:22:44 | 2021-12-20 21:15:57 | 2021-12-20 |           17 |          0 |            8 |            9 |
| ****** |  20870 | 2021-10-20 05:22:06 | 2021-12-20 20:45:08 | 2021-12-20 |            8 |          0 |            3 |            5 |
| ****** |  20934 | 2021-05-11 07:01:38 | 2021-12-20 21:00:57 | 2021-12-20 |           11 |          0 |            3 |            8 |
| ****** |  21009 | 2021-05-17 08:09:46 | 2021-12-20 20:54:07 | 2021-12-20 |           47 |          0 |           22 |           25 |
```

**Reply by JFen10** [reply]
Thank you for the detailed update @fabd . Appreciate all your efforts in maintaining the site and I am understanding that these things can happen sometimes. I appreciate your quick resolution on what is a free service.

I think in my case as the easiest solution is to move the 221 cards all to the fail pile we can go with that, and then I can start them again from there. I've been feeling a bit rusty on much of them lately anyway so this isn't too much of an issue for me to put them back there. Please let me know once I'll be able to check again once this has been completed.

**Reply by c4b3z0n** [reply]
Thanks for the update!
You can move mine to failed, I had only one correct so it's no problem 😜 
As a coder myself I can relate, sometimes this happens.
And thanks for mantaining the site, it has been essential for learning kanji!

**Reply by fabd** [reply]
@c4b3z0n Your flashcards should be back in order, and you should be able to use the site as usual. 19 cards should appear in your red pile and Flashcard List : 享 処 加 嗅 奮 恒 拒 是 末 村 混 玩 男 赴 辻 造 銃 錮 題

**Reply by fabd** [reply]
@JFen10 Your flashcards have also been corrected. Cf. sent email.

**Comment by fabd** [comment]
## Concerning the six remaining users who haven't got in touch

**Issue:  Not being able to see the SRS chart, not able to start a SRS review.**

For these 6 remaining users affected by the bug , I have moved these cards to the Restudy List, in order to resolve the issue quicker (taking in account there number of cards affected for these 6 users is between 12-40 cards).

I will mark this bug as fixed -- there is still a PR to be merged that adds further validation to prevent any similar issue in the future.
